### PR TITLE
fix: break on outer call and not on nested func

### LINKED ISSUE #124: Breakpoint maps to *nested* function instead of *outer* call
## Steps to reproduce
```c#
async Task<int> Method()
{
    ParallelOptions parallelOptions = new() { MaxDegreeOfParallelism = 3 };
    var userHandlers = new []{"a", "b","c","d"};

    await Parallel.ForEachAsync(userHandlers, parallelOptions, async (uri, token) =>
    {
        await new HttpClient().GetAsync("https://google.com");
    });

    return 0;
}

int result = await Method();

```

Add a breakpoint on line 6:`await Parallel.ForEachAsync(userHandlers, parallelOptions, async (uri, token) =>`

* actual behavior - when debugging, the breakpoing actually changes to be line 7, and stopped on each iteration of the callback.
* expected behavior to me would be stopping once on the outer call, and when 'resume'ing, move to `return 0;` below.

## Mitigation
I have found that this patch resolves this, but it drops support from what the comment says here- 
```patch
diff --git a/src/metadata/modules_sources.cpp b/src/metadata/modules_sources.cpp
index ef4ba4e..a735ceb 100644
--- a/src/metadata/modules_sources.cpp
+++ b/src/metadata/modules_sources.cpp
@@ -177,21 +177,20 @@ namespace
                 break;
             }
             // result was found on previous cycle, check for closest nested method
             // need it in case of breakpoint setuped at lines without code and before nested method, for example:
             // {
             //  <-- breakpoint at line without code (inside method)
             //     void Method() {...}
             // }
             else if (result && lineNum <= (*lower).startLine && (*lower).endLine <= result->endLine)
             {
-                closestNestedToken = (*lower).methodDef;
                 break;
             }
             else
                 break;
         }
 
         if (result)
         {
             auto find = multiMethodBpData.find(*result);
             if (find != multiMethodBpData.end()) // only constructors segments could be part of multiple methods
```

## Proper solution?
Maybe on managed part? on `SymbolReader.cs` there's this condition - https://github.com/Samsung/netcoredbg/blob/db69338cf1606d8d327de0eaaf1694d8463af135/src/managed/SymbolReader.cs#L795

I added some prints -

```c#
if (nestedToken != 0)
{
    SequencePoint nested_p = FirstSequencePointForSourceLine(ref reader, nestedToken);
    StreamWriter sw = new StreamWriter("/tmp/managed.log");
    sw.WriteLine("current_p - " + current_p.StartLine + ":" + current_p.StartColumn + " until " + current_p.EndLine + ":" + current_p.EndColumn);
    sw.WriteLine("nested_p - " + nested_p.StartLine + ":" + nested_p.StartColumn + " until " + nested_p.EndLine + ":" + nested_p.EndColumn);
    sw.Close();
    if (current_p.EndLine > nested_p.EndLine || (current_p.EndLine == nested_p.EndLine && current_p.EndColumn > nested_p.EndColumn))
    {
        list.Add(new resolved_bp_t(nested_p.StartLine, nested_p.EndLine, nested_p.Offset, nestedToken));
        // (tokenNum > 1) can have only lines, that added to multiple constructors, in this case - we will have same for all Tokens,
        // we need unique tokens only for breakpoints, prevent adding nestedToken multiple times.
        break;
    }
}
nestedToken = 0; // Don't check nested block next cycle (will have same results).

list.Add(new resolved_bp_t(current_p.StartLine, current_p.EndLine, current_p.Offset, methodToken));
```

and they are -

```
current_p - 9:5 until 12:8
nested_p - 10:5 until 10:6
```

I'm not sure if the condition needs to be changed, or it's given some bad input. That's because the nested function is not really between `10:5` to `10:6`, and actaully contains multiple lines. So I'm not sure how to really solve this properly.

### PR DESCRIPTION
**Pull request by noam-sol**
Fixes #124

I tried to follow @viewizard advice on https://github.com/Samsung/netcoredbg/issues/124#issuecomment-1591178401 and it works nicely for the use-case described there.

I have run the test suite and they all passed (except for `VSCodeTestEvalArraysIndexers` which seems to be flaky for me on master branch anyway).

Would love to hear your thoughts, thanks!

### CODE REVIEW

File: src/managed/SymbolReader.cs [hunk 1]
```diff
@@ -767,12 +771,16 @@ SequencePoint FirstSequencePointForSourceLine(ref MetadataReader reader, int met
 
                         if (p.EndLine != nearestSP.EndLine)
                         {
-                            if (p.EndLine < nearestSP.EndLine)
+                            if (reqPos == Position.First && p.EndLine < nearestSP.EndLine)
+                                nearestSP = p;
+                            else if (reqPos == Position.Last && p.EndLine > nearestSP.EndLine)
```
Comment (viewizard) [inline src/managed/SymbolReader.cs:776]: We probably don't need 2 branches for `if` statement here, just use 1 with modified condition:
```
    if ((a) ||
        (b))
         nearestSP = p;
```

File: src/managed/SymbolReader.cs [hunk 2]
```diff
@@ -767,12 +771,16 @@ SequencePoint FirstSequencePointForSourceLine(ref MetadataReader reader, int met
...
                                 nearestSP = p;
                         }
                         else
                         {
-                            if (p.EndColumn < nearestSP.EndColumn)
+                            if (reqPos == Position.First && p.EndColumn < nearestSP.EndColumn)
+                                nearestSP = p;
+                            else if (reqPos == Position.Last && p.EndColumn > nearestSP.EndColumn)
```
Comment (viewizard) [inline src/managed/SymbolReader.cs:783]: Same here.

File: src/managed/SymbolReader.cs [hunk 3]
```diff
@@ -788,16 +796,31 @@ SequencePoint FirstSequencePointForSourceLine(ref MetadataReader reader, int met
...
+                        //     {
+                        //        await new HttpClient().GetAsync("https://google.com");
+                        //     });
+                        // nesetedToken here is the annonymous async func, and having a breakpoing at the 1st line should
+                        // break on the outer call.
+                        SequencePoint nested_start_p = SequencePointForSourceLine(Position.First, ref reader, nestedToken);
+                        SequencePoint nested_end_p = SequencePointForSourceLine(Position.Last, ref reader, nestedToken);
+                        if (nested_start_p.StartLine > current_p.StartLine || (nested_start_p.StartLine == current_p.StartLine && nested_start_p.StartColumn > current_p.StartColumn)
+                             && nested_end_p.EndLine < current_p.EndLine || (nested_end_p.EndLine == current_p.EndLine && nested_end_p.EndColumn < current_p.EndColumn )
+                        ) {
```
Comment (viewizard) [inline src/managed/SymbolReader.cs:816]: This condition looks wrong for me
```
    if ( a1>b1 || ( a1==b1 && c1>d1 ) && a2<b2 || ( a2==b2 && c2<d2 ) )
```
probably, should be
```
    if ( ( a1>b1 || ( a1==b1 && c1>d1 ) ) && ( a2<b2 || ( a2==b2 && c2<d2 ) ) )
```
instead?

**Review by viewizard — CHANGES_REQUESTED** [review]

**Review by viewizard — COMMENTED** [review]
Looks good for me. Let me some time, I will put your commit into our internal dev repo for CI testing and internal reviewing.

**Comment by noam-sol** [comment]
@viewizard thanks! let me know if there's anything I can help out with.

**Comment by gbalykov** [comment]
Fix is part of latest release

**Pull request closed without merging**